libraries/chain/controller.cpp

```diff
@@ -2707,8 +2718,28 @@ struct controller_impl {
 
          if( s == controller::block_status::incomplete ) {
             log_irreversible();
-            pacemaker->beat();
          }
+
+         // TODO: temp transition to instant-finality, happens immediately after block with new_finalizer_policy
+         auto transition = [&](auto& fork_db, auto& head) -> bool {
+            const auto& bsp = std::get<std::decay_t<decltype(head)>>(cb.bsp);
+            std::optional<block_header_extension> ext = bsp->block->extract_header_extension(instant_finality_extension::extension_id());
+            if (ext) {
+               const auto& if_extension = std::get<instant_finality_extension>(*ext);
+               if (if_extension.new_finalizer_policy) {
+                  ilog("Transition to instant finality happening after block ${b}", ("b", bsp->block_num()));
+                  hs_irreversible_block_num = bsp->block_num();
+
+                  log_irreversible();
```

Comment:
Will `log_irreversible` be executed twice as there was one above?

Reply:
It will be executed twice. But `if_irreversible_block_num` is set before the second call. `log_irreversible` determines internally if it should do anything. Always safe to call it multiple times.